compiler/ir/backend.jvm/src/org/jetbrains/kotlin/backend/jvm/codegen/FunctionCodegen.kt

```diff
@@ -51,17 +50,22 @@ open class FunctionCodegen(
         var methodVisitor = createMethod(flags, signature)
 
         val hasSyntheticFlag = flags.and(Opcodes.ACC_SYNTHETIC) != 0
-        generateParameterNames(irFunction, methodVisitor, signature, state, hasSyntheticFlag)
+        if (state.generateParametersMetadata && !hasSyntheticFlag) {
+            generateParameterNames(irFunction, methodVisitor, signature, state)
+        }
+
+        if (irFunction.origin != IrDeclarationOrigin.FUNCTION_FOR_DEFAULT_PARAMETER) {
```

Comment:
I wonder if we should use the same condition to generate parameter annotations (below)? I really hope that there's no case in the current backend where we generate method annotations, but not parameter annotations, or vice versa. (Note BTW that `$default` method was exactly this case before 1.3.40 where we deliberately decided not to generate any annotations on it: [KT-30289](https://youtrack.jetbrains.com/issue/KT-30289).)

Reply:
There shouldn't be, but there is one more special case that we need to handle: When we regenerate anonymous objects in the inliner, the constructor should not have parameter annotations, since we can end up with fewer parameters after inlining (see, e.g., the lambda class corresponding to `{ s() }()` in `testKt17091`).

I suppose we should fix this in the inliner instead, but the code is sufficiently complex that I want to think a bit more about it before making any changes.

Reply:
Right, anonymous object copy should take a different code path since it's being done via bytecode. I don't mind postponing this change though. But it'd be nice to update the TODO below if possible, just to prevent someone else spending time researching why specific (and different) origins are handled here and below

Reply:
Sure, I've added a comment to that effect.